lib/service/storage/virtualIndex.ts

```diff
@@ -0,0 +1,200 @@
+  }
+
+  public virtualIndexMap: Map<string, string> = new Map(); //Key : virtual index, value : real index
+
+  async init(): Promise<void> {
+    await this.buildCollection();
+    await this.initVirtualIndexList();
+
+    global.kuzzle.onAsk(VirtualIndex.createEvent, (physical, virtual) =>
+      this.addInSoftIndexMap(physical, virtual)
+    );
+
+    global.kuzzle.onAsk(VirtualIndex.deleteEvent, (info) =>
+      this.removeInSoftIndexMap(info)
+    );
+  }
+
+  addInSoftIndexMap(physical, virtual) {
+    this.virtualIndexMap.set(virtual, physical);
+  }
+
+  removeInSoftIndexMap(notification) {
+    this.virtualIndexMap.delete(notification);
+  }
+
+  getRealIndex(name: string): string {
+    if (this.virtualIndexMap.has(name)) {
+      return this.virtualIndexMap.get(name);
+    }
+    return name;
+  }
+
+  isVirtual(name: string): boolean {
+    return this.virtualIndexMap.has(name);
+  }
+
+  getId(index: string, id: string): string {
+    if (this.isVirtual(index)) {
+      return index + id;
+    }
+    return id;
+  }
+
+  randomString(size = 20) {
+    const buf = Buffer.alloc(size);
+    for (let i = 0; i < 20; i++) {
+      buf.writeUInt8((Math.random() * 255) & 255, i);
+    }
+    return buf.toString("base64").slice(0, size);
+  }
+
+  getVirtualId(index: string, id: string): string {
+    if (this.isVirtual(index) && id.startsWith(index)) {
+      return id.substring(index.length, id.length);
+    }
+    return id;
+  }
+
+  public async createVirtualIndex(virtualIndex: string, index: string) {
+    global.kuzzle.emit(VirtualIndex.createVirtualIndexEvent, {
+      physical: index,
+      virtual: virtualIndex,
+    });
+
+    this.virtualIndexMap.set(virtualIndex, index);
+    await global.kuzzle.ask(
+      "core:storage:private:document:create",
+      "virtual-indexes",
+      "list",
+      { physical: index, virtual: virtualIndex },
+      { id: index + virtualIndex }
+    );
+  }
+
+  async removeVirtualIndex(index: string) {
+    const realIndex = this.virtualIndexMap.get(index);
+    global.kuzzle.emit(VirtualIndex.deleteVirtualIndexEvent, {
+      virtual: index,
+    });
+    this.virtualIndexMap.delete(index);
+    const id = realIndex + index;
+    await global.kuzzle.ask(
+      "core:storage:private:document:delete",
+      "virtual-indexes",
+      "list",
+      id
+    );
+  }
+
+  async initVirtualIndexList() {
+    if (this.virtualIndexMap.size === 0) {
+      return;
+    }
+    this.virtualIndexMap = new Map<string, string>();
+    let from = 0;
+    let total;
+
+    do {
+      const list = await global.kuzzle.ask(
+        "core:storage:private:document:search",
+        "virtual-indexes",
+        "list",
+        { from: from, size: 100 }
+      );
+      total = list.total;
+      for (const hit of list.hits) {
+        this.virtualIndexMap.set(hit._source.virtual, hit._source.physical);
+      }
+      from += 100;
+    } while (from < total);
+  }
+
+  async buildCollection() {
```

Comment:
Rather than creating a new internal index, you should create your collection in the internal `kuzzle` index

Reply:
Done